### PR TITLE
feat(artifact_collector): artifacts collection for conditional steps

### PR DESCRIPTION
**Pull request by miltolstoy**
This PR contains:
1. Handling conditional step branches artifacts
2. Conditional steps tests update to check files created instead of the log line

Similar changes will be applied to the `configuration.report_artifacts` in the next PR after this change will be reviewed and updated accordingly.

Relates #709

### CODE REVIEW

File: universum/modules/artifact_collector.py [hunk 1]
```diff
@@ -128,48 +139,61 @@ def preprocess_artifact_list(self, artifact_list, ignore_already_existing=False)
...
         self.html_output.artifact_dir_ready = True
         artifact_list = []
         report_artifact_list = []
         for configuration in project_configs.all():
             if configuration.artifacts:
-                path = utils.parse_path(configuration.artifacts, self.settings.project_root)
-                artifact_list.append(dict(path=path, clean=configuration.artifact_prebuild_clean))
+                artifact_list.append(self.get_config_artifact(configuration))
+            if configuration.is_conditional:
+                artifacts_info = self.get_conditional_step_branches_artifacts(configuration)
```
Comment (i-keliukh) [inline universum/modules/artifact_collector.py:194]: I think you could implement the artifact support by adding artifacts of both branches to the list. All other code should not depend on the steps structure and execution status when dealing with artifacts.
Fortunately or unfortunately this is how artifacts are currently processed (if I am not mistaken).

File: universum/modules/artifact_collector.py [hunk 2]
```diff
@@ -184,6 +208,32 @@ def set_and_clean_artifacts(self, project_configs: Configuration, ignore_existin
                                                                     name, True, report_artifact_list,
                                                                     ignore_existing_artifacts)
 
+    def get_conditional_step_branches_artifacts(self, configuration):
+        """
+        Both branches have artifacts -> return ConditionalStepArtifacts(dict, dict)
+        One branch have artifact -> return dict
```
Comment (i-keliukh) [inline universum/modules/artifact_collector.py:214]: have -> has

File: universum/modules/artifact_collector.py [hunk 3]
```diff
@@ -228,8 +278,30 @@ def collect_artifacts(self):
...
+    def collect_conditional_step_artifacts(self, artifacts_info):
+        succeeded_artifact_path = artifacts_info.succeeded_config_artifact["path"]
+        succeeded_artifact_exists = os.path.exists(succeeded_artifact_path)
+        failed_artifact_path = artifacts_info.failed_config_artifact["path"]
+        failed_artifact_exists = os.path.exists(failed_artifact_path)
+
+        # At this point we don't know which branch step was executed,
+        # so assuming that any single branch artifact presence is good
+        if succeeded_artifact_exists and failed_artifact_exists:
+            raise CriticalCiException("Both conditional step branches artifacts exist")
```
Comment (i-keliukh) [inline universum/modules/artifact_collector.py:296]: I think this logic is incorrect.
Generally we allow the same artifact paths to be specified multiple times in different steps. 
If some regular step has artifact path "a" (and creates corresponding artifact) and skipped conditional step has the same artifact path "a", this line would trigger the exception, which is incorrect.

**Comment by i-keliukh** [comment]
The requirements for the artifact processing for conditional steps are missing in this PR.
I think it would be good to have them listed in the form of product documentation included in this PR.

**Comment by k-dovgan** [comment]
Please ignore this comment.

File: universum/modules/artifact_collector.py (hunk 3 above)
Comment (miltolstoy) [reply universum/modules/artifact_collector.py:297]: This code became redundant after the artifacts collection moment was changed from the end of the build to the end of each step.
Removed.

File: tests/test_conditional_steps.py [hunk 4]
```diff
@@ -25,16 +27,34 @@ def check_conditional_step_success(tmpdir, capsys, conditional_step_passed):
...
+
     config = inspect.cleandoc(f'''
         from universum.configuration_support import Configuration, Step
 
-        true_branch_step = Step(name='{true_branch_step_name}', command=['touch', '{true_branch_step_name}'])
-        false_branch_step = Step(name='{false_branch_step_name}', command=['touch', '{false_branch_step_name}'])
-        conditional_step = Configuration([dict(name='conditional',
-            command=['bash', '-c', 'exit {conditional_step_exit_code}'],
-            if_succeeded=true_branch_step, if_failed=false_branch_step)])
+        true_branch_step = Step(**{str(true_branch_step)})
```
Comment (i-keliukh) [inline tests/test_conditional_steps.py:48]: It looks like this variable is not needed. But I think you can keep it for readability.

File: tests/test_conditional_steps.py [hunk 5]
```diff
@@ -56,11 +107,20 @@ def check_conditional_step(tmpdir, capsys, config_file, conditional_step_passed)
...
+
+    assert os.path.exists(os.path.join(artifacts_dir, conditional_step_artifact))
+
+    expected_artifact = true_branch_step_artifact if is_conditional_step_passed else false_branch_step_artifact
+    if expected_artifact:
+        assert os.path.exists(os.path.join(artifacts_dir, expected_artifact))
+
+    unexpected_artifact = false_branch_step_artifact if is_conditional_step_passed else true_branch_step_artifact
+    if unexpected_artifact:
+        assert not os.path.exists(os.path.join(artifacts_dir, unexpected_artifact))
```
Comment (i-keliukh) [inline tests/test_conditional_steps.py:126]: No empty line at the end of file

File: tests/test_conditional_steps.py [hunk 6]
```diff
@@ -1,40 +1,89 @@
...
+    conditional_step_artifact = steps_info.conditional_step["artifacts"]
+    steps_info.true_branch_step.command = ["touch", conditional_step_artifact]
+    steps_info.true_branch_step.artifacts = conditional_step_artifact
 
-def build_config_file(tmpdir, conditional_step_passed):
-    conditional_step_exit_code = 0 if conditional_step_passed else 1
+    check_conditional_step(tmpdir, capsys, steps_info)
 
+
+def test_no_artifact_in_executed_branch_step(tmpdir, capsys):
```
Comment (i-keliukh) [inline tests/test_conditional_steps.py:39]: Is the logic of this test to check that artifacts are not collected if they are not specified for step?
How does it relate to conditional steps? 
If it doesn't, I think this is a test for basic functionality of artifacts collection and should be a part of its tests.
If the idea is to test that artifacts are not collected for the opposing branch (fail for successful case and success for failure case), then the test is incorrect.

File: universum/modules/artifact_collector.py [hunk 7]
```diff
@@ -175,6 +182,28 @@ def set_and_clean_artifacts(self, project_configs: Configuration, ignore_existin
             with self.structure.block(block_name=name, pass_errors=True):
                 self.preprocess_artifact_list(report_artifact_list, ignore_existing_artifacts)
 
+    def get_conditional_step_branches_artifacts(self, configuration: Step) -> List[ArtifactInfo]:
```
Comment (i-keliukh) [inline universum/modules/artifact_collector.py:185]: Misleading parameter name: configuration for type Step.

File: universum/modules/artifact_collector.py [hunk 8]
```diff
@@ -175,6 +182,28 @@ def set_and_clean_artifacts(self, project_configs: Configuration, ignore_existin
...
+        succeeded_config_artifact: Optional[ArtifactInfo] = self.get_config_artifact_if_exists(configuration.if_succeeded)
+        failed_config_artifact: Optional[ArtifactInfo] = self.get_config_artifact_if_exists(configuration.if_failed)
+
+        if succeeded_config_artifact and failed_config_artifact:
```
Comment (i-keliukh) [inline universum/modules/artifact_collector.py:189]: Please take a look a this code:
```python
result = []
if succeeded_config_artifact: result.append(succeeded_config_artifact)
if failed_config_artifact: result.append(failed_config_artifact)
return result
```

What do you think about its readability?

Comment (i-keliukh) [reply universum/modules/artifact_collector.py:189]: Alternatively, the entire function can be merged into `preprocess_artifact_list` by using similar logic

**Comment by miltolstoy** [comment]
> The requirements for the artifact processing for conditional steps are missing in this PR. I think it would be good to have them listed in the form of product documentation included in this PR.

It appears that artifact processing for conditional steps is similar to regular steps. It makes more sense to describe conditional steps logic in general and make small notes about artifacts processing in several places.
Agreed to be done in a separate PR because it's not related to the artifacts processing only.

File: universum/modules/artifact_collector.py (hunk 8 above)
Comment (miltolstoy) [reply universum/modules/artifact_collector.py:189]: > What do you think about its readability?

Much better, thanks.)

**Review by i-keliukh — APPROVED** [review]

**Pull request merged**